Added support for COMPUTED values inside CSV files for loadData change

zbynekvavros opened the pull request:
Based on https://forum.liquibase.org/#Topic/49382000002091061.

nvoxland commented:
Looks like a good change. We'll be catching up on master-based pull requests soon. Thanks

zbynekvavros commented:
Any idea when you gonna to pull requests?

ro-rah commented:
Hi @zbynekvavros 

Thanks for your pull request!  And your patience.

Here’s what happens next:

A member of the Liquibase team will take a look at your contribution and may suggest 
changes, additional tests, or provide other feedback. The PR will be prioritized 
according to our internal development and testing capacity.

We’ll let you know when it’s ready to move to the next step or if any changes are needed.

Thanks,

Ronak

ro-rah commented:
Hi @zbynekvavros,

Would you be willing to write a [unit test](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests) for this PR?

Looks like we could use this file being added:
liquibase-core/src/test/java/liquibase/statement/ExecutablePreparedStatementBaseTest.java

This would be an incredible help in getting this PR into Conditioning here internally.

Thanks for the work already done!

-Ronak

zbynekvavros commented:
Hey,

well I can do unit test for sure!

Just not sure I understand, you want me to add this file liquibase-core/src/test/java/liquibase/statement/ExecutablePreparedStatementBaseTest.java? I don't see it in current [https://github.com/liquibase/liquibase/tree/master/liquibase-core/src/test/java/liquibase/statement](https://github.com/liquibase/liquibase/tree/master/liquibase-core/src/test/java/liquibase/statement).

Please let me know.

Thanks,
Zbynek

ro-rah commented:
Awesome @zbynekvavros !  Yes, this would be a newly added file, and that's cool b/c you will be the first one to increase the testing in this area!

biergit commented:
Note that this is not a backward compatible solution to https://liquibase.jira.com/browse/CORE-3208
Before this bug was introduced `SOME_SEQUENCE.nextval;1;1` was valid CSV data to import to Oracle.
With https://github.com/liquibase/liquibase/pull/944/commits/3df26067ffd15d3605ee6d48e8778af832301953 `SOME_SEQUENCE.nextval` would be executed as statement which is not valid and would throw an exception.

zbynekvavros commented:
I'm not sure I understand, as you can see from my change, the 'valueComputed' was never used at all. So there isn't anything that can be broken.
The issues you mentioned was opened April 13, 2018, 9:02 PM and is still open. Is there any plan to make PR of this?

biergit commented:
Sorry, basically I just wanted to point out that while `SOME_SEQUENCE.nextval;1;1` was resolved to `NULL` it will result in an exception with your change.

zbynekvavros commented:
I'm quite new in liquibase development (has been only user for quite some time) so maybe I am missing something but you say *it will result in an exception*. Does this mean it is actually in use somewhere/by somebody? After my change you would have to use `select SOME_SEQUENCE.nextval from DUAL` instead of `SOME_SEQUENCE.nextval` but that's it, isn't it?

biergit commented:
I'm currently updating from `3.5.3` and I'm facing the `NULL` problem and came here in hope of finding a fix for it. The sequence reference is classified as computed value but it will still not be resolved correctly as it has been in version `3.5.3`. I am no liquibase expert so I am not sure if migrating all my `loadData` to `select SOME_SEQUENCE.nextval from DUAL` is really the way to. If it is then I guess it should be documented somewhere.

zbynekvavros commented:
Well I am no expert either but as you can see in my [post ](https://forum.liquibase.org/#Topic/49382000002091061) on forums the computed value **works fine** using an [insert](https://www.liquibase.org/documentation/changes/insert.html) 

`
<changeSet author="liquibase-docs" id="insert-example">
    <insert catalogName="cat"
            dbms="h2, oracle"
            schemaName="public"
            tableName="person">
        <column name="address" valueComputed="SELECT...."/>
    </insert>
</changeSet>
`

but **is being completely ignored** using [loadData](https://www.liquibase.org/documentation/changes/load_update_data.html) that uses CSV files

`
<changeSet author="liquibase-docs" id="loadUpdateData-example">
    <loadUpdateData catalogName="cat"
            commentLineStartsWith="A String"
            encoding="UTF-8"
            file="com/example/users.csv"
            onlyUpdate="true"
            primaryKey="pk_id"
            quotchar="'"
            relativeToChangelogFile="true"
            schemaName="public"
            separator=","
            tableName="person"
            usePreparedStatements="true">
        <column name="address" type="varchar(255)"/>
        <column name="person_id" type="computed"/>
    </loadUpdateData>
</changeSet>
`

and users.csv

`address_column_content,SELECT id FROM person WHERE ...`

My change is merely making sure this `computed` value is working for CSV files as well.

zbynekvavros commented:
Well I am no expert either but as you can see in my [post ](https://forum.liquibase.org/#Topic/49382000002091061) on forums the computed value **works fine** using an [insert](https://www.liquibase.org/documentation/changes/insert.html) 

```
<changeSet author="liquibase-docs" id="insert-example">
    <insert catalogName="cat"
            dbms="h2, oracle"
            schemaName="public"
            tableName="person">
        <column name="address" valueComputed="SELECT...."/>
    </insert>
</changeSet>
```

but **is being completely ignored** using [loadData](https://www.liquibase.org/documentation/changes/load_update_data.html) that uses CSV files

```
<changeSet author="liquibase-docs" id="loadUpdateData-example">
    <loadUpdateData catalogName="cat"
            commentLineStartsWith="A String"
            encoding="UTF-8"
            file="com/example/users.csv"
            onlyUpdate="true"
            primaryKey="pk_id"
            quotchar="'"
            relativeToChangelogFile="true"
            schemaName="public"
            separator=","
            tableName="person"
            usePreparedStatements="true">
        <column name="address" type="varchar(255)"/>
        <column name="person_id" type="computed"/>
    </loadUpdateData>
</changeSet>
```

and users.csv

`address_column_content,SELECT id FROM person WHERE ...`

My change is merely making sure this `computed` value is working for CSV files as well.

biergit commented:
You are definitely moving in the right direction. I hope https://liquibase.jira.com/browse/CORE-3208 will be addressed soon too :)

zbynekvavros commented:
Well the best way possible is doing a PR :)

biergit commented:
I see a lot of open PRs here.
Regarding your change wouldn't it be better if the computer parameter was not determined by an extra roundtrip to the DB but rather passed in as simple value to the query. So (?,?) - > (SOME_SEQUENCE.nextval,?)
Excuse me being brief here I am on mobile.

ro-rah commented:
Hey Guys, 

There are a lot of PRs open at the moment but we at Liquibase are working on it!  Our next couple of sprints should be taking in community PRs.  The strategy is 2 parts: 1) Process backlog of PRs and 2) Make it easier to submit and accept PRs.

For the part 1, that will take time but, not too much.  There is additional focus and hours being spent here in the next few sprints.
For part 2, we have created some documentation on:
- [How to submit a Liquibase PR](https://liquibase.jira.com/wiki/spaces/LB/pages/1274904896/How+to+Contribute+Code+to+Liquibase+Core)
- Creating [unit tests](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
- Creating [integration tests](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests).

Thanks for collaborating, I'm looking to process this and change the status to Conditioning soon (our entry into our engineering process).

-Ronak

ro-rah commented:
Sized this as at least Medium because:
- I am no expert but looks like we don't have consensus on the PR method of solving the issue (would like a review internally).
- There seems to be a bigger issue, opportunity to address https://liquibase.jira.com/browse/CORE-3208

biergit commented:
Hey @ro-rah thank you for the quick and thorough heads-up. I would love to see the focus on fixing the regressions and preventing them in the future with better tests 🙂

zbynekvavros commented:
Thanks for stepping in @ro-rah. I completely agree there has to be consensus. As I see it, I could update my change so that the database function is indeed set as statement parameter to avoid hitting DB one more time. But I still think the computed value should support complete subquery - this would bring more flexibility. I'll try to have a look tomorrow.

ro-rah commented:
@zbynekvavros , thanks for the update!  I personally love seeing users like you and @biergit interacting, just here to help!

zbynekvavros commented:
Okay so what should be the next step? @ro-rah you said you would like review internally, should we wait for that?
@biergit I did tried to use the computed value without additional trip do DB but simply setting computed value as parameter using for example [setObject()](https://docs.oracle.com/javase/7/docs/api/java/sql/PreparedStatement.html#setObject(int,%20java.lang.Object))  fills column with the SQL, not the value it represents as any complex objects are forbidden in parameters. Any hints?

After all I must stress out that this fix is only for [loadData](https://www.liquibase.org/documentation/changes/load_data.html) change, for [loadUpdateData](https://www.liquibase.org/documentation/changes/load_update_data.html) it works fine already (we switched to this).

ro-rah commented:
It's in our court to process. We've jump started our PR process but have a backlog of about 130+ PRs to go through.

We'll provide feedback if needed and merge, best estimate now is it will make it in the current quarter. 

Thanks for collaborating!

biergit commented:
Hey @zbynekvavros I suggest you take a look at the patch supplied at https://liquibase.jira.com/browse/CORE-3208
Setting the computed value by using `setObject` does not work as for example a value `SOME_SEQUENCE.nextval` will just be interpreted as String.

zbynekvavros commented:
Oh silly me, thanks for the pointer @biergit :)

So what I did is

- updated my PR so that there is no extra DB hit for computed values, instead the computed values are directly included in SQL as in [https://liquibase.jira.com/browse/CORE-3208](https://liquibase.jira.com/browse/CORE-3208)
- added tests for all affected classes
- rebased my fork to current

Now I tried to write tests using [Mockito](https://site.mockito.org/) but as seen above, with Mockito included some groovy tests are failing.

@ro-rah what should be the proper way to solve this? I see @nvoxland commented out all tests using mocks in ae83cc7f8ee34a272c1786b1a3b8a8118c7fcbae. Maybe there is some greater idea about using mocks? Or I could remove/comment out my tests as well...

ro-rah commented:
I will bring this up with @nvoxland.  Thanks for writing up the issue @zbynekvavros !

b-gyula commented:
It seems to work for _ONLY ONE computed value for ALL ROWS_ in the given column, 
not for different values in different rows like:

<table>
<tr>
	<td>id
	<td>comp
<tr>
	<td>1
	<td>id+1
<tr>
	<td>2
	<td>id-1
</table>

MalloD12 commented:
Approved.

**Review and testing results**:

As part of the review which overall looks good to me I have performed below actions:

-  As it was a pretty old PR (almost 3 years) I've resolved some merge conflicts for which we made a mix of what was implemented on this PR and what we implemented in the last time.
- Tested this change manually on **postgresql** (v14.4).
- Added an integration test so this change can be easily tested across the different supported DBs. Basically what this test does is perform a **loadData** with a CSV file which contains a call to a value that needs to be computed in one of its columns, plus also perform some insert/update to test these types of values are processed successfully.

**Things to be aware of**:
- This change has an impact only on this type of value (computed).

**Things to worry about**:
- Nothing

filipelautert reviewed (approved):
Besides it's a big PR, the core changes are small. Added a lot of tests.

biergit commented:
Hey @filipelautert and @MalloD12 could you confirm that https://github.com/liquibase/liquibase/pull/944#issuecomment-618939540 works now?

MalloD12 commented:
> Hey @filipelautert and @MalloD12 could you confirm that [#944 (comment)](https://github.com/liquibase/liquibase/pull/944#issuecomment-618939540) works now?

Hi @biergit, 

This is working in oracle too, we have tried running a changeset like the one below and it worked without issues:
```tes_col,test_col2
1,seq_test.NEXTVAL
2,seq_test.NEXTVAL 
```
Please let us know if you any other question.

Thanks,
Daniel.

XDelphiGrl commented on liquibase-integration-tests/src/test/resources/changelogs/common/common.tests.changelog.xml:
```diff
@@ -1177,4 +1177,22 @@
+                  file="computedValueData.csv"
+                  relativeToChangelogFile="true">
+            <column name="test_col" type="COMPUTED"/>
+        </loadData>
```
@filipelautert , do you think it is worthwhile to add a test for `loadUpdateData`? 

[https://docs.liquibase.com/change-types/load-update-data.html](https://docs.liquibase.com/change-types/load-update-data.html)

MalloD12 replied:
@XDelphiGrl I think it would be pretty much the same than the changeset added for loadData, but it's not a big deal adding an extra changeset for testing loadUpdateData. So, I'm adding it now.

marchof commented:
The new URL of the original discussion is https://forum.liquibase.org/t/loaddata-csv-computedvalue/4035/6
